central/postgres/pruning.go

```diff
@@ -32,14 +32,28 @@ const (
 		(SELECT 1 FROM clusters WHERE nodes.clusterid = clusters.Id)`
 
 	// Explain Analyze indicated that 2 statements for PLOP is faster than one.
-	deleteOrphanedPLOPDeployments = `DELETE FROM listening_endpoints WHERE processindicatorid in (SELECT id from process_indicators pi WHERE NOT EXISTS
+	deleteOrphanedPLOPDeployments = `DELETE FROM listening_endpoints WHERE poduid IS NULL AND processindicatorid in (SELECT id from process_indicators pi WHERE NOT EXISTS
 		(SELECT 1 FROM deployments WHERE pi.deploymentid = deployments.Id) AND 
 		(signal_time < now() at time zone 'utc' - INTERVAL '%d MINUTES' OR signal_time is NULL))`
 
-	deleteOrphanedPLOPPods = `DELETE FROM listening_endpoints WHERE processindicatorid in (SELECT id from process_indicators pi WHERE NOT EXISTS
+	deleteOrphanedPLOPPods = `DELETE FROM listening_endpoints WHERE poduid IS NULL AND processindicatorid in (SELECT id from process_indicators pi WHERE NOT EXISTS
 		(SELECT 1 FROM pods WHERE pi.poduid = pods.Id) AND 
 		(signal_time < now() at time zone 'utc' - INTERVAL '%d MINUTES' OR signal_time is NULL))`
 
+	// Unfortunately if a listening endpoint is marked as being open there is no indication of how old it is.
+	// This leads to a possible race condition where a listening endpoint reaches the database before the deployment,
+	// and the pruning job happens to run before the deployment information arrives in the database.
+	// This should be rare, so this should be acceptable. This could be improved by adding a timestamp to the listening endpoints table
+	deleteOrphanedPLOPDeploymentsWithPodUID = `DELETE FROM listening_endpoints WHERE poduid IS NOT NULL AND NOT EXISTS
+		(SELECT 1 FROM deployments WHERE listening_endpoints.deploymentid = deployments.Id)`
+
+	// Unfortunately if a listening endpoint is marked as being open there is no indication of how old it is.
+	// This leads to a possible race condition where a listening endpoint reaches the database before the pod,
+	// and the pruning job happens to run before the pod information arrives in the database.
+	// This should be rare, so this should be acceptable. This could be improved by adding a timestamp to the listening endpoints table
+	deleteOrphanedPLOPPodsWithPodUID = `DELETE FROM listening_endpoints WHERE poduid IS NOT NULL AND NOT EXISTS
```

Comment:
does the `poduid IS NOT NULL` provide value here?  I suspect postgres would optimize this knowing that it is comparing it to a field that cannot be null.

Reply:
`poduid` can be null. Check here https://github.com/stackrox/stackrox/pull/7989

I did some upgrade tests. This resulted in null values for poduid for the listening endpoints that existed before the upgrade. E.g
```
select count(*) from listening_endpoints where poduid is null;
-[ RECORD 1 ]
count | 83

select count(*) from listening_endpoints where poduid is not null;
-[ RECORD 1 ]
count | 0
```

I made this check for poduid being null based on @md2119 comment about treating old listening endpoints the old way and treating the new listening endpoints the new way.

Reply:
@JoukoVirtanen yes the `poduid` can be null here.  That wasn't my point.  If it is null, then this 
`(SELECT 1 FROM pods WHERE listening_endpoints.poduid = pods.Id)`
should return false.  So I contend (without testing or analyzing) that the statement is the same whether you have `poduid IS NOT NULL` in it or not.

Reply:
I added the explain plan in the description. I also made the change locally and ran the unit tests. Some of the unit tests failed

```
=== RUN   TestPruning/TestRemoveOrphanedPLOPs
=== RUN   TestPruning/TestRemoveOrphanedPLOPs/no_deployments_nor_pods_-_remove_plops_with_PodUid_since_there_are_no_pods
=== NAME  TestPruning/TestRemoveOrphanedPLOPs
    pruning_test.go:553: 
        	Error Trace:	/home/jvirtane/go/src/github.com/stackrox/stackrox/central/postgres/pruning_test.go:553
        	Error:      	Not equal: 
        	            	expected: 3
        	            	actual  : 0
        	Test:       	TestPruning/TestRemoveOrphanedPLOPs
=== RUN   TestPruning/TestRemoveOrphanedPLOPs/deployments_one_missing_pod_-_remove_plops_with_PodUid_with_no_matching_pod_even_though_there_are_deployments
=== NAME  TestPruning/TestRemoveOrphanedPLOPs
    pruning_test.go:553: 
        	Error Trace:	/home/jvirtane/go/src/github.com/stackrox/stackrox/central/postgres/pruning_test.go:553
        	Error:      	Not equal: 
        	            	expected: 5
        	            	actual  : 2
        	Test:       	TestPruning/TestRemoveOrphanedPLOPs
=== RUN   TestPruning/TestRemoveOrphanedPLOPs/one_missing_deployments_no_missing_pods_-_remove_plops_with_PodUid_with_no_matching_deployments_even_though_there_are_matching_pods
=== NAME  TestPruning/TestRemoveOrphanedPLOPs
    pruning_test.go:553: 
        	Error Trace:	/home/jvirtane/go/src/github.com/stackrox/stackrox/central/postgres/pruning_test.go:553
        	Error:      	Not equal: 
        	            	expected: 5
        	            	actual  : 2
        	Test:       	TestPruning/TestRemoveOrphanedPLOPs
=== RUN   TestPruning/TestRemoveOrphanedPLOPs/no_missing_deployments_or_pods_but_plops_are_expired_-_remove_all_expired_plops
```

For example in this test case
```
                {
                        name: "no deployments nor pods - remove plops with PodUid since there are no pods",
                        initialPlops: []*storage.ProcessListeningOnPortStorage{
                                fixtures.GetPlopStorage1(),
                                fixtures.GetPlopStorage2(),
                                fixtures.GetPlopStorage3(),
                                fixtures.GetPlopStorage4(),
                                fixtures.GetPlopStorage5(),
                                fixtures.GetPlopStorage6(),
                        },
                        deployments:           set.NewFrozenStringSet(),
                        pods:                  set.NewFrozenStringSet(),
                        expectedPlopDeletions: []string{fixtureconsts.PlopUID4, fixtureconsts.PlopUID5, fixtureconsts.PlopUID6},
                },
```
There are no deployments or pods, so 
```
FROM listening_endpoints WHERE NOT EXISTS (SELECT 1 FROM deployments WHERE listening_endpoints.deploymentid = deployments.Id);
```
matches all listening endpoints. Then the three without PodUids are filtered out by `IS NOT NULL`. So, without `IS NOT NULL` all of the listening endpoints are deleted and with it only the ones with PodUids are deleted.

Reply:
To clarify the test above GetPlopStorage4 through GetPlopStorage6 have PodUid, but GetPlopStorage1 through GetPlopStorage3. Therefore in the above test GetPlopStorage4 through GetPlopStorage6 should be removed and the others should not be removed.